File: edition4/js/ch02/2-5.js
function Node(data, next) {
  this.data = data;
  this.next = next;
};

function findLoopBeginning(head) {
  var n1 = head;
  var n2 = head;

  while (n2.next !== null) {
    n1 = n1.next;
    n2 = n2.next.next;

    if (n1.data === n2.data) {
      break;
    }
  }

  if (!n2.next) {
    return null;
  }

  n1 = head;
  while (n1 !== n2) {
    n1 = n1.next;
    n2 = n2.next;
  }

  return n1;
}

// tests
var assert = require('assert');

describe('findLoopBeginning', function() {

  it('should fail when there is no loop', function() {
    var n2 = new Node('3', null);
    var n1 = new Node('2', n2);
    var head = new Node('1', n1);

    var actual = findLoopBeginning(head);
    assert.equal(actual, null);
  });

  it('should succeed and return the start when there is a loop', function() {
    var n9 = new Node('3', null);
    var n8 = new Node('2', n9);
    var n7 = new Node('1', n8);
    var n6 = new Node('3', n7);
    var n5 = new Node('2', n6);
    var n4 = new Node('1', n5);
    var n3 = new Node('3', n4);
    var n2 = new Node('2', n3);
    var n1 = new Node('1', n2);
    var head = new Node('0', n1);
    n9.next = n4;

    var actual = findLoopBeginning(head);
    assert.equal(actual, n4);
  });

});

File: edition4/c/ch02/2-5.cpp
#include <assert.h>

#include <algorithm>
#include <cmath>
#include <iostream>
#include <string>
#include <set>

using namespace std;

class Node {
  public:
    int value;
    Node *next;
    Node(int value, Node *next);
};

Node::Node(int value, Node *next) {
  this->value = value;
  this->next = next;
}

class LinkedList {
  public:
    Node *head;
    LinkedList();
    void add(int value);
    void deleteFromNode(Node *n);
    int length();
    void print();
    Node *findLoopBeginning();
};

LinkedList::LinkedList() {
  this->head = NULL;
}

void LinkedList::add(int value) {
  if (!this->head) {
    this->head = new Node(value, NULL);
  } else {
    Node *temp = this->head;

    while (temp->next) {
      temp = temp->next;
    }

    temp->next = new Node(value, NULL);
  }
}

int LinkedList::length() {
  int length = 0;
  Node *curr = this->head;

  while (curr) {
    length++;
    curr = curr->next;
  }

  return length;
}

void LinkedList::print() {
  Node *curr = this->head;

  while (curr) {
    cout << curr->value << " -> ";
    curr = curr->next;
  }

  cout << "NULL" << endl;
}

Node* LinkedList::findLoopBeginning() {
  set<int> seen;
  Node *curr = this->head;

  while (curr && seen.find(curr->value) == seen.end()) {
    seen.insert(curr->value);
    curr = curr->next;
  }

  return curr;
}

int main() {
  // tests
  LinkedList l;

  assert(l.findLoopBeginning() == NULL);

  l.add(1);
  l.add(2);
  l.add(3);
  l.add(4);
  l.add(2);
  Node* n = l.findLoopBeginning();
  assert(n->value == 2);

  return 0;
}
